src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewDataEvaluator.cs

```diff
@@ -98,57 +99,58 @@ private static ViewDataInfo EvalComplexExpression(object indexableObject, string
+                if (targetInfo != null)
                 {
-                    if (string.IsNullOrEmpty(postExpression))
+                    if (leftExpression.Length == expression.Length)
                     {
-                        return subTargetInfo;
+                        // Nothing remaining in expression after leftExpression.
+                        return targetInfo;
                     }
 
-                    if (subTargetInfo.Value != null)
+                    if (targetInfo.Value != null)
                     {
-                        var potential = InnerEvalComplexExpression(subTargetInfo.Value, postExpression);
-                        if (potential != null)
+                        var rightExpression = expression.Substring(leftExpression.Length + 1);
+                        targetInfo = InnerEvalComplexExpression(targetInfo.Value, rightExpression);
+                        if (targetInfo != null)
                         {
-                            return potential;
+                            return targetInfo;
                         }
                     }
                 }
+
+                leftExpression = GetNextShorterExpression(leftExpression);
             }
+            while (!string.IsNullOrEmpty(leftExpression));
 
             return null;
         }
 
-        // Produces an enumeration of combinations of property names given a complex expression in the following order:
-        //  this["one.two.three.four"]
-        //  this["one.two.three][four"]
-        //  this["one.two][three.four"]
-        //  this["one][two.three.four"]
+        // Given "one.two.three.four" initially, calls return
+        //  "one.two.three"
+        //  "one.two"
+        //  "one"
+        //  ""
         // Recursion of InnerEvalComplexExpression() further sub-divides these cases to cover the full set of
         // combinations shown in Eval(ViewDataDictionary, string) comments.
-        private static IEnumerable<ExpressionPair> GetRightToLeftExpressions(string expression)
+        private static string GetNextShorterExpression(string expression)
         {
-            yield return new ExpressionPair(expression, string.Empty);
+            if (string.IsNullOrEmpty(expression))
+            {
+                return string.Empty;
+            }
 
             var lastDot = expression.LastIndexOf('.');
-
-            var subExpression = expression;
-            var postExpression = string.Empty;
-
-            while (lastDot > -1)
+            if (lastDot == -1)
             {
-                subExpression = expression.Substring(0, lastDot);
-                postExpression = expression.Substring(lastDot + 1);
-                yield return new ExpressionPair(subExpression, postExpression);
-
-                lastDot = subExpression.LastIndexOf('.');
+                return string.Empty;
             }
+
+            return expression.Substring(startIndex: 0, length: lastDot);
```

Comment:
Considered that but it means every failed check of every key would run through the keys in the dictionary and / or the properties in a POCO. The approach also mandates extensions or alternates to `TryGetValueProvider`, another `public` class in this namespace.